SiameseDataset.py

```diff
@@ -0,0 +1,271 @@
+import collections
+import h5py
+import numpy.random as random
+import numpy as np
+
+from CachedDataset2 import CachedDataset2
+from HDFDataset import FeatureSequenceStreamParser, SparseStreamParser, SegmentAlignmentStreamParser
+from Dataset import DatasetSeq
+
+
+
+class SiameseHDFDataset(CachedDataset2):
+  """
+  SiameseHDFDataset class allows to do sequence sampling for weakly-supervised training
+  It accepts data in the format of NextGenHDFDataset and performs sampling of sequence triplets before each epoch
+  Here we assume that each dataset from the input .hdf has a single label
+  Sampled triplets are splitted into different streams
+  In the config we can access streams by e.g. ["data:features_0", "data:features_1", "data:features_2"],
+  "data:features_0" stands for features of all sequences from the input data,
+  "data:features_1" is a stream of pairs, where for each sequence from "data:features_0" SiameseHDFDataset randomly samples a sequence with the same label
+  "data:features_2" is a stream, where for each sequence from "data:features_0" SiameseHDFDataset randomly samples a sequence with a different label
+  targets are splitted into different streams as well, e.g. "data:classes_0", "data:classes_1", "data:classes_2"
+
+  SiameseHDFDataset also supports non-uniform sampling and accepts a path to .npz matrix
+  Rows of this matrix should have probabilities for each of the classes to be sampled
+  This probability distribution might reflect class similarities
+
+  This dataset might be useful for metric learning,
+  where we want to learn such representations of input sequences, that those which belong to the same class are close together,
+  while those with different labels should have representations far away from each other
+  """
+  parsers = { 'feature_sequence'  : FeatureSequenceStreamParser,
+              'sparse'            : SparseStreamParser,
+              'segment_alignment' : SegmentAlignmentStreamParser }
+
+  def __init__(self, input_stream_name, seq_label_stream='words', class_distribution=None, files=None, *args, **kwargs):
+    """
+
+    :param str input_stream_name: name of a feature stream
+    :param str seq_label_stream: name of a stream with labels
+    :param str class_distribution: path to .npz file of size n x n (n is a number of classes),
+           where each line i contains probs of other classes to be picked in triplets
+           when sampling a pair for element from class i
+    :param files: list of paths to .hdf files
+    :param args: dict[str]
+    :param kwargs: dict[str]
+    """
+    super(SiameseHDFDataset, self).__init__(*args, **kwargs)
+    self.input_stream_name = input_stream_name
+    if class_distribution is not None:
+      self.class_probs = np.load(class_distribution)['arr_0']
+    else:
+      self.class_probs = None
+
+    self.files           = []
+    self.h5_files        = []
+    self.all_seq_names   = [] # all_seq_names[(int)seq_index] = (string) sequence_name
```

Comment:
There should be two spaces between the end of the command (`[]`) and `#`.
Use PyCharm, it will notify you about standard Python conventions / code formatting standards.